### DISCUSSION TITLE
Option to specify 'bin-directory' analog to 'contents-directory'

### DISCUSSION THREAD

**Discussion by adkrprog** [pyinstaller/pyinstaller#8834] [Features & Development]
Hi

The possibility to specify --contents-directory is a very nice thing to have multiple software in parallel. Could you implement the same concept for the binary part (--bin-directory)? 
The use case is to use GNU stow for versioning binary files and not to have contents-directory in the search path.

actual:
``` 
|---myapp
|---contents-directory
      |---some libs
``` 

new:
``` 
|---bin-directory
      |---myapp
|---contents-directory
      |---some libs
``` 

I would very appreciate this extension to pyinstaller.

Thanks, Adi

**Comment by rokm** [comment]
> ```
> actual:
> |---myapp
> |---contents-directory
> |---some libs
> ```

It's actually just

```
actual:
|---myapp
|---contents-directory
```

because *everything* goes into contents-directory (or not, if you set it to '.' to have the old onedir layout).

> ```
> new:
> |---bin-directory
> |---myapp
> |---contents-directory
> |---some libs
> ```

I'm not sure what you want to put into `bin-directory`? Most python packages assume the layout where binaries (package-bundled shared libs, extensions, even package-bundled binaries) are located next to package data (and source .py files, if they need to be collected). So even if we were to go about splitting those apart (like we have to for example in macOS .app bundles), we'd need to symlink them back together for things to work. So I don't think this would help with "not having contents-directory in the search path".

**Reply by adkrprog** [reply]
I intend to put the 'bootloader' (the file you start the app) into the e.g. bin directory
e.g.:
pyinstaller --contents-directory=lib --bin-directory=bin myscript.py

bin/myscript
lib/(all the other stuff)

instead of:
myscript
lib/(all the other stuff)

**Reply by rokm** [reply]
Ah yeah, I see now that lack of quote tags in your original post removed the indentation whitespaces, which made the tree look flat.

**Reply by rokm** [reply]
This basically means that you want contents-directory path to be allowed to point outside the directory in which the executable is located (i.e., `../lib`). Which we are currently explicitly forbidding because it would place the contents directory outside of the `dist` directory during the build process. Having an option to specify sub-directory for executable (either explicitly via `--bin-directory` or `--executable-directory`; or perhaps even by allowing `name=bin/myscript` in EXE) would indeed alleviate that.

But you could also work around the current limitation by setting contents-directory to `AAAlib` for the build, then rename it to `lib`, move executable into `bin` directory and use hex editor to change `AAAlib` into `../lib`.

**Reply by adkrprog** [reply]
That would allow me to implement a clean UNIX like setup. I plan to setup a ci/cd pipeline to build the software and distribute it over NFS to the clients. Therefore the workaround is not an option.
If I can help, please let me know. 
Many thanks, Adi

**Reply by bwoodsend** [reply]
When we implemented `--contents-directory`, we really really didn't want it to escalate into anything that would require more permutative testing. Even `--contents-directory=.` was a very reluctant concession just to stop people from whinging at us.

**Reply by bwoodsend** [reply]
I still don't see how a `bin` directory solves your problem.

**Reply by adkrprog** [reply]
I plan do deploy a large number of applications. All packages are then linked with gnu stow with the bin directory added to the $PATH. Having all ‚lib‘ stuff in the $PATH creates unnecessary overhad. Therefore I would like to have that separated, like unix does.

**Reply by rokm** [reply]
Why can't you just link `/my/custom/location/program1/program1` to `/usr/local/bin/program1` and be done with it? Does `stow` mandate this `bin`/`lib` separation (i.e., you point it to directory and with current layout it links both `program1` and its `_internal` directory)? If so, you could also just create a `bin` directory that contains symlink to `../program1` and point `stow` to that? (EDIT: hmm, or actually not, if you need to point it at `/my/custom/location/program1` directory - in that case it would end up linking `bin`, `_internal`, and `program1`)
